## ECBM E4040 Homework 4 - Programming Problem

In [1]:
import theano
import hw4a
import hw4b

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


### Problem a: Recurrent Neural Network
#### Bullet 1: Experiment with RNNSLU

In [2]:
from hw4a import test_rnnslu
test_rnnslu(**{
        'fold': 3,
        'lr': 0.0970806646812754,
        'verbose': True,
        'decay': True,
        'win': 7,
        'nhidden': 200,
        'seed': 345,
        'emb_dimension': 50,
        'nepochs': 60,
        'savemodel': False,
        'normal': True,
        'folder':'../result'})
#size of context window: 7
#dimension of word embedding: 50

verbose: True
normal: True
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 68.18 best test F1 66.22
NEW BEST: epoch 1 valid F1 80.63 best test F1 77.98
NEW BEST: epoch 2 valid F1 83.79 best test F1 80.66
NEW BEST: epoch 3 valid F1 88.42 best test F1 84.92
NEW BEST: epoch 4 valid F1 89.65 best test F1 85.97
NEW BEST: epoch 5 valid F1 91.02 best test F1 86.87
NEW BEST: epoch 6 valid F1 91.45 best test F1 88.81
NEW BEST: epoch 7 valid F1 92.05 best test F1 89.89
NEW BEST: epoch 8 valid F1 93.78 best test F1 90.69
NEW BEST: epoch 9 valid F1 94.19 best test F1 91.27
NEW BEST: epoch 10 valid F1 94.7 best test F1 91.64
NEW BEST: epoch 11 valid F1 94.88 best test F1 92.23
NEW BEST: epoch 12 valid F1 95.4 best test F1 92.73
NEW BEST: epoch 13 valid F1 95.45 best test F1 92.79

NEW BEST: epoch 15 valid F1 95.55 best test F1 

(<hw4_nn.RNNSLU at 0x7f38e23e8f10>,
 {'labels2idx': {'B-aircraft_code': 0,
   'B-airline_code': 1,
   'B-airline_name': 2,
   'B-airport_code': 3,
   'B-airport_name': 4,
   'B-arrive_date.date_relative': 5,
   'B-arrive_date.day_name': 6,
   'B-arrive_date.day_number': 7,
   'B-arrive_date.month_name': 8,
   'B-arrive_date.today_relative': 9,
   'B-arrive_time.end_time': 10,
   'B-arrive_time.period_mod': 11,
   'B-arrive_time.period_of_day': 12,
   'B-arrive_time.start_time': 13,
   'B-arrive_time.time': 14,
   'B-arrive_time.time_relative': 15,
   'B-booking_class': 16,
   'B-city_name': 17,
   'B-class_type': 18,
   'B-compartment': 19,
   'B-connect': 20,
   'B-cost_relative': 21,
   'B-day_name': 22,
   'B-day_number': 23,
   'B-days_code': 24,
   'B-depart_date.date_relative': 25,
   'B-depart_date.day_name': 26,
   'B-depart_date.day_number': 27,
   'B-depart_date.month_name': 28,
   'B-depart_date.today_relative': 29,
   'B-depart_date.year': 30,
   'B-depart_time.end_time': 3

In [1]:
from hw4a import test_rnnslu
test_rnnslu(**{
        'fold': 3,
        'lr': 0.0970806646812754,
        'verbose': True,
        'decay': True,
        'win': 5,
        'nhidden': 200,
        'seed': 345,
        'emb_dimension': 50,
        'nepochs': 60,
        'savemodel': False,
        'normal': True,
        'folder':'../result'})
#size of context window: 5
#dimension of word embedding: 50

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


verbose: True
normal: True
win: 5
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 69.47 best test F1 65.79
NEW BEST: epoch 1 valid F1 80.18 best test F1 76.07
NEW BEST: epoch 2 valid F1 82.58 best test F1 79.6
NEW BEST: epoch 3 valid F1 87.23 best test F1 83.35
NEW BEST: epoch 4 valid F1 89.3 best test F1 85.58
NEW BEST: epoch 5 valid F1 90.55 best test F1 87.2

NEW BEST: epoch 7 valid F1 91.56 best test F1 89.85
NEW BEST: epoch 8 valid F1 93.18 best test F1 91.09
NEW BEST: epoch 9 valid F1 93.7 best test F1 91.58
NEW BEST: epoch 10 valid F1 94.3 best test F1 91.64

NEW BEST: epoch 12 valid F1 94.86 best test F1 92.29
NEW BEST: epoch 13 valid F1 94.87 best test F1 92.42
NEW BEST: epoch 14 valid F1 95.4 best test F1 92.36


NEW BEST: epoch 17 valid F1 95.91 best test F1 93.41
NEW BEST: epoch 18 valid F1 95.96 best test F1 

(<hw4_nn.RNNSLU at 0x7fa2e5d6a850>,
 {'labels2idx': {'B-aircraft_code': 0,
   'B-airline_code': 1,
   'B-airline_name': 2,
   'B-airport_code': 3,
   'B-airport_name': 4,
   'B-arrive_date.date_relative': 5,
   'B-arrive_date.day_name': 6,
   'B-arrive_date.day_number': 7,
   'B-arrive_date.month_name': 8,
   'B-arrive_date.today_relative': 9,
   'B-arrive_time.end_time': 10,
   'B-arrive_time.period_mod': 11,
   'B-arrive_time.period_of_day': 12,
   'B-arrive_time.start_time': 13,
   'B-arrive_time.time': 14,
   'B-arrive_time.time_relative': 15,
   'B-booking_class': 16,
   'B-city_name': 17,
   'B-class_type': 18,
   'B-compartment': 19,
   'B-connect': 20,
   'B-cost_relative': 21,
   'B-day_name': 22,
   'B-day_number': 23,
   'B-days_code': 24,
   'B-depart_date.date_relative': 25,
   'B-depart_date.day_name': 26,
   'B-depart_date.day_number': 27,
   'B-depart_date.month_name': 28,
   'B-depart_date.today_relative': 29,
   'B-depart_date.year': 30,
   'B-depart_time.end_time': 3

In [ ]:
from hw4a import test_rnnslu
test_rnnslu(**{
        'fold': 3,
        'lr': 0.0970806646812754,
        'verbose': True,
        'decay': True,
        'win': 5,
        'nhidden': 200,
        'seed': 345,
        'emb_dimension': 70,
        'nepochs': 60,
        'savemodel': False,
        'normal': True,
        'folder':'../result'})
#size of context window: 5
#dimension of word embedding: 70

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


verbose: True
normal: True
win: 5
savemodel: False
fold: 3
seed: 345
emb_dimension: 70
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 71.08 best test F1 68.73
NEW BEST: epoch 1 valid F1 81.18 best test F1 78.84
NEW BEST: epoch 2 valid F1 84.19 best test F1 80.02
NEW BEST: epoch 3 valid F1 87.51 best test F1 83.83
NEW BEST: epoch 4 valid F1 89.67 best test F1 86.15
NEW BEST: epoch 5 valid F1 90.61 best test F1 87.03
NEW BEST: epoch 6 valid F1 90.76 best test F1 88.0
NEW BEST: epoch 7 valid F1 92.14 best test F1 89.27
NEW BEST: epoch 8 valid F1 93.56 best test F1 90.83
NEW BEST: epoch 9 valid F1 94.41 best test F1 92.32
NEW BEST: epoch 10 valid F1 94.81 best test F1 92.74
NEW BEST: epoch 11 valid F1 94.84 best test F1 93.09
NEW BEST: epoch 12 valid F1 95.62 best test F1 92.97



NEW BEST: epoch 16 valid F1 95.65 best test F1 93.27
NEW BEST: epoch 17 valid F1 96.06 best test 

(<hw4_nn.RNNSLU at 0x7f99d00beb90>,
 {'labels2idx': {'B-aircraft_code': 0,
   'B-airline_code': 1,
   'B-airline_name': 2,
   'B-airport_code': 3,
   'B-airport_name': 4,
   'B-arrive_date.date_relative': 5,
   'B-arrive_date.day_name': 6,
   'B-arrive_date.day_number': 7,
   'B-arrive_date.month_name': 8,
   'B-arrive_date.today_relative': 9,
   'B-arrive_time.end_time': 10,
   'B-arrive_time.period_mod': 11,
   'B-arrive_time.period_of_day': 12,
   'B-arrive_time.start_time': 13,
   'B-arrive_time.time': 14,
   'B-arrive_time.time_relative': 15,
   'B-booking_class': 16,
   'B-city_name': 17,
   'B-class_type': 18,
   'B-compartment': 19,
   'B-connect': 20,
   'B-cost_relative': 21,
   'B-day_name': 22,
   'B-day_number': 23,
   'B-days_code': 24,
   'B-depart_date.date_relative': 25,
   'B-depart_date.day_name': 26,
   'B-depart_date.day_number': 27,
   'B-depart_date.month_name': 28,
   'B-depart_date.today_relative': 29,
   'B-depart_date.year': 30,
   'B-depart_time.end_time': 3

In [2]:
from hw4a import test_rnnslu
test_rnnslu(**{
        'fold': 3,
        'lr': 0.0970806646812754,
        'verbose': True,
        'decay': True,
        'win': 7,
        'nhidden': 200,
        'seed': 345,
        'emb_dimension': 70,
        'nepochs': 60,
        'savemodel': False,
        'normal': True,
        'folder':'../result'})
#size of context window: 7
#dimension of word embedding: 70

verbose: True
normal: True
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 70
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 70.07 best test F1 66.15
NEW BEST: epoch 1 valid F1 80.96 best test F1 77.41
NEW BEST: epoch 2 valid F1 83.99 best test F1 80.12
NEW BEST: epoch 3 valid F1 88.62 best test F1 84.49
NEW BEST: epoch 4 valid F1 90.17 best test F1 86.41
NEW BEST: epoch 5 valid F1 91.23 best test F1 87.17
NEW BEST: epoch 6 valid F1 92.41 best test F1 89.12
NEW BEST: epoch 7 valid F1 93.22 best test F1 90.43
NEW BEST: epoch 8 valid F1 94.11 best test F1 91.24
NEW BEST: epoch 9 valid F1 94.31 best test F1 90.99
NEW BEST: epoch 10 valid F1 94.73 best test F1 91.37
NEW BEST: epoch 11 valid F1 94.96 best test F1 92.3
NEW BEST: epoch 12 valid F1 95.56 best test F1 92.51
NEW BEST: epoch 13 valid F1 95.8 best test F1 92.44

NEW BEST: epoch 15 valid F1 96.11 best test F1 

(<hw4_nn.RNNSLU at 0x7f09db9815d0>,
 {'labels2idx': {'B-aircraft_code': 0,
   'B-airline_code': 1,
   'B-airline_name': 2,
   'B-airport_code': 3,
   'B-airport_name': 4,
   'B-arrive_date.date_relative': 5,
   'B-arrive_date.day_name': 6,
   'B-arrive_date.day_number': 7,
   'B-arrive_date.month_name': 8,
   'B-arrive_date.today_relative': 9,
   'B-arrive_time.end_time': 10,
   'B-arrive_time.period_mod': 11,
   'B-arrive_time.period_of_day': 12,
   'B-arrive_time.start_time': 13,
   'B-arrive_time.time': 14,
   'B-arrive_time.time_relative': 15,
   'B-booking_class': 16,
   'B-city_name': 17,
   'B-class_type': 18,
   'B-compartment': 19,
   'B-connect': 20,
   'B-cost_relative': 21,
   'B-day_name': 22,
   'B-day_number': 23,
   'B-days_code': 24,
   'B-depart_date.date_relative': 25,
   'B-depart_date.day_name': 26,
   'B-depart_date.day_number': 27,
   'B-depart_date.month_name': 28,
   'B-depart_date.today_relative': 29,
   'B-depart_date.year': 30,
   'B-depart_time.end_time': 3

#### I experiment the size of context window with 5 and 7. And experiment the dimension of word embeddings with 50 and 70. The result is shown as below:

#### context window is 7, dimension of word embeddings is 50 : epoch 57 valid F1 97.23 best test F1 94.2
#### context window is 5, dimension of word embeddings is 50 : epoch 57 valid F1 96.96 best test F1 93.57
#### context window is 5, dimension of word embeddings is 70 : epoch 49 valid F1 97.23 best test F1 93.95
#### context window is 7, dimension of word embeddings is 70 : epoch 27 valid F1 96.96 best test F1 93.83

#### When the size of context window is 7 and the dimension of word embedding is 50, we can achieve the best test accuracy.

#### Bullet 2: Experiment with RNNSLU without normalizing word embeddings

In [1]:
from hw4a import test_rnnslu
test_rnnslu()
#size of context window: 7
#dimension of word embedding: 50
#without normalizing the word embeddings after each update

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


verbose: True
normal: False
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 68.96 best test F1 63.35
NEW BEST: epoch 1 valid F1 81.28 best test F1 76.9
NEW BEST: epoch 2 valid F1 84.73 best test F1 81.46
NEW BEST: epoch 3 valid F1 88.04 best test F1 84.06
NEW BEST: epoch 4 valid F1 90.05 best test F1 86.02
NEW BEST: epoch 5 valid F1 91.31 best test F1 86.93
NEW BEST: epoch 6 valid F1 92.65 best test F1 88.75
NEW BEST: epoch 7 valid F1 93.21 best test F1 89.79
NEW BEST: epoch 8 valid F1 94.0 best test F1 90.6
NEW BEST: epoch 9 valid F1 94.86 best test F1 90.68
NEW BEST: epoch 10 valid F1 95.13 best test F1 91.98

NEW BEST: epoch 12 valid F1 95.68 best test F1 92.92
NEW BEST: epoch 13 valid F1 95.91 best test F1 92.79

NEW BEST: epoch 15 valid F1 96.25 best test F1 93.48
NEW BEST: epoch 16 valid F1 96.33 best test F1

(<hw4_nn.RNNSLU at 0x7f47f02867d0>,
 {'labels2idx': {'B-aircraft_code': 0,
   'B-airline_code': 1,
   'B-airline_name': 2,
   'B-airport_code': 3,
   'B-airport_name': 4,
   'B-arrive_date.date_relative': 5,
   'B-arrive_date.day_name': 6,
   'B-arrive_date.day_number': 7,
   'B-arrive_date.month_name': 8,
   'B-arrive_date.today_relative': 9,
   'B-arrive_time.end_time': 10,
   'B-arrive_time.period_mod': 11,
   'B-arrive_time.period_of_day': 12,
   'B-arrive_time.start_time': 13,
   'B-arrive_time.time': 14,
   'B-arrive_time.time_relative': 15,
   'B-booking_class': 16,
   'B-city_name': 17,
   'B-class_type': 18,
   'B-compartment': 19,
   'B-connect': 20,
   'B-cost_relative': 21,
   'B-day_name': 22,
   'B-day_number': 23,
   'B-days_code': 24,
   'B-depart_date.date_relative': 25,
   'B-depart_date.day_name': 26,
   'B-depart_date.day_number': 27,
   'B-depart_date.month_name': 28,
   'B-depart_date.today_relative': 29,
   'B-depart_date.year': 30,
   'B-depart_time.end_time': 3

#### The result is epoch 31 valid F1 97.56 best test F1 94.0, the test accuracy is similar to the result when experiment with normalizing word embeddings

#### Bullet 3: Implement RNN with two hidden layers

In [1]:
from hw4a import test_rnnslu2
test_rnnslu2()

Using gpu device 0: Tesla K80 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


verbose: True
nhidden2: 100
win: 7
nhidden1: 200
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 56.73 best test F1 51.6
NEW BEST: epoch 1 valid F1 67.46 best test F1 64.99
NEW BEST: epoch 2 valid F1 74.88 best test F1 72.25
NEW BEST: epoch 3 valid F1 80.21 best test F1 77.82
NEW BEST: epoch 4 valid F1 83.96 best test F1 80.87
NEW BEST: epoch 5 valid F1 85.87 best test F1 82.86
NEW BEST: epoch 6 valid F1 86.75 best test F1 83.75
NEW BEST: epoch 7 valid F1 88.91 best test F1 86.27
NEW BEST: epoch 8 valid F1 89.32 best test F1 86.9
NEW BEST: epoch 9 valid F1 90.59 best test F1 88.48
NEW BEST: epoch 10 valid F1 90.77 best test F1 88.74
NEW BEST: epoch 11 valid F1 91.55 best test F1 89.15
NEW BEST: epoch 12 valid F1 92.27 best test F1 89.44
NEW BEST: epoch 13 valid F1 93.45 best test F1 90.65



NEW BEST: epoch 17 valid F1 93.88 best test

#### The result is epoch 57 valid F1 96.3 best test F1 92.75, the test accuracy is slightly lower than the experiment with just one hidden layer.

#### Bullet 4: Experiment with layer normalization

In [1]:
from hw4a import test_rnnslu
test_rnnslu()

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


verbose: True
normal: True
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 68.27 best test F1 64.49
NEW BEST: epoch 1 valid F1 82.63 best test F1 79.4
NEW BEST: epoch 2 valid F1 87.14 best test F1 84.07
NEW BEST: epoch 3 valid F1 90.79 best test F1 87.94

NEW BEST: epoch 5 valid F1 93.75 best test F1 90.08

NEW BEST: epoch 7 valid F1 94.62 best test F1 92.33
NEW BEST: epoch 8 valid F1 95.25 best test F1 92.16
NEW BEST: epoch 9 valid F1 95.69 best test F1 92.87

NEW BEST: epoch 11 valid F1 96.19 best test F1 93.78
NEW BEST: epoch 12 valid F1 96.38 best test F1 93.66


NEW BEST: epoch 15 valid F1 96.63 best test F1 93.87

NEW BEST: epoch 17 valid F1 96.69 best test F1 93.8
NEW BEST: epoch 18 valid F1 96.72 best test F1 94.05

NEW BEST: epoch 20 valid F1 96.81 best test F1 93.96























BEST RESULT: epo

(<hw4a.RNNSLU at 0x7f66fd6f3a10>,
 {'labels2idx': {'B-aircraft_code': 0,
   'B-airline_code': 1,
   'B-airline_name': 2,
   'B-airport_code': 3,
   'B-airport_name': 4,
   'B-arrive_date.date_relative': 5,
   'B-arrive_date.day_name': 6,
   'B-arrive_date.day_number': 7,
   'B-arrive_date.month_name': 8,
   'B-arrive_date.today_relative': 9,
   'B-arrive_time.end_time': 10,
   'B-arrive_time.period_mod': 11,
   'B-arrive_time.period_of_day': 12,
   'B-arrive_time.start_time': 13,
   'B-arrive_time.time': 14,
   'B-arrive_time.time_relative': 15,
   'B-booking_class': 16,
   'B-city_name': 17,
   'B-class_type': 18,
   'B-compartment': 19,
   'B-connect': 20,
   'B-cost_relative': 21,
   'B-day_name': 22,
   'B-day_number': 23,
   'B-days_code': 24,
   'B-depart_date.date_relative': 25,
   'B-depart_date.day_name': 26,
   'B-depart_date.day_number': 27,
   'B-depart_date.month_name': 28,
   'B-depart_date.today_relative': 29,
   'B-depart_date.year': 30,
   'B-depart_time.end_time': 31,

#### The result is epoch 20 valid F1 96.81 best test F1 93.96, the test accuracy is slightly lower than the result without layer normalization. However, when we experiment without layer normalization it reach the best result on epoch 57. With layer normalization it reach the best result on epoch 20. We have a faster convergence when we add layer normalization.

#### Bullet 5: Observation on embedding pairs

In [1]:
from hw4a import test_rnnslu
emb, idx2label, idx2word = test_rnnslu(**{
        'fold': 3,
        'lr': 0.0970806646812754,
        'verbose': True,
        'decay': True,
        'win': 7,
        'nhidden': 200,
        'seed': 345,
        'emb_dimension': 50,
        'nepochs': 60,
        'savemodel': False,
        'normal': True,
        'folder':'../result'})

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


verbose: True
normal: True
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 68.27 best test F1 64.49
NEW BEST: epoch 1 valid F1 82.63 best test F1 79.4
NEW BEST: epoch 2 valid F1 87.14 best test F1 84.07
NEW BEST: epoch 3 valid F1 90.79 best test F1 87.94

NEW BEST: epoch 5 valid F1 93.75 best test F1 90.08

NEW BEST: epoch 7 valid F1 94.62 best test F1 92.33
NEW BEST: epoch 8 valid F1 95.25 best test F1 92.16
NEW BEST: epoch 9 valid F1 95.69 best test F1 92.87

NEW BEST: epoch 11 valid F1 96.19 best test F1 93.78
NEW BEST: epoch 12 valid F1 96.38 best test F1 93.66


NEW BEST: epoch 15 valid F1 96.63 best test F1 93.87

NEW BEST: epoch 17 valid F1 96.69 best test F1 93.8
NEW BEST: epoch 18 valid F1 96.72 best test F1 94.05

NEW BEST: epoch 20 valid F1 96.81 best test F1 93.96























BEST RESULT: epo

In [2]:
from hw4a import test_rnnslu
emb, idx2label, idx2word = test_rnnslu(**{
        'fold': 3,
        'lr': 0.0970806646812754,
        'verbose': False,
        'decay': True,
        'win': 7,
        'nhidden': 200,
        'seed': 345,
        'emb_dimension': 50,
        'nepochs': 60,
        'savemodel': False,
        'normal': True,
        'folder':'../result'})

... loading the dataset
... building the model
... training
BEST RESULT: epoch 20 valid F1 96.81 best test F1 93.96 with the model ../result


In [22]:
from scipy import spatial
import numpy as np
emb = np.asarray(emb)


words = {}
for i, v in idx2word.iteritems():
    if v == 'to':
        words[v] = emb[i]
    if 'UNK' not in v and 'DIGIT' not in v and "'" not in v and len(v) > 4:
        words[v] = emb[i]

word_dict = {}
for w1 in words.keys():
    for w2 in words.keys():
        k = w1 + '-' + w2
        if w1 != w2 and k not in word_dict:
            word_dict[k] = words[w1] - words[w2]
            


print(spatial.distance.cosine(word_dict['sunday-people'], word_dict['people-costs']))
print(spatial.distance.cosine(word_dict['flying-seattle'], word_dict['flying-minneapolis']))
print(spatial.distance.cosine(word_dict['expensive-costs'], word_dict['maximum-costs']))
print(spatial.distance.cosine(word_dict['sunday-requesting'], word_dict['friday-requesting']))
print(spatial.distance.cosine(word_dict['sunday-requesting'], word_dict['friday-requesting']))
print(spatial.distance.cosine(word_dict['daily-aircraft'], word_dict['california-aircraft']))

1.35302921438
0.264798183497
0.407927241336
0.0754148778254
0.0754148778254
0.402767076123


#### As we can see from above, when the word pairs are irrelevant such as "sunday/people" with "people/costs", the distance between the two vector are large. When the words paris very similar such as "sunday/requesting" with "friday/requesting", the distance between the two vector are very small. When the word pairs have some kind of relation for example "flying/seattle" with "flying/minneapolis", the distance between the two vector are relatively small.

### Problem b: Learning Parity Function
#### Bullet 1: Shallow MLP

#### 8-bit input

In [8]:
from hw4b import test_mlp_parity
test_mlp_parity(learning_rate=0.1, n_bit=8,n_hiddenLayers=2,L1_reg=0.00, L2_reg=0.0001, n_epochs=1500, batch_size=100, n_hidden=10)

... building the model
...training
training @ iter =  0
epoch 1, minibatch 10/10, validation error 47.400000 %
     epoch 1, minibatch 10/10, test error of best model 55.000000 %
epoch 2, minibatch 10/10, validation error 47.600000 %
epoch 3, minibatch 10/10, validation error 46.400000 %
     epoch 3, minibatch 10/10, test error of best model 56.000000 %
epoch 4, minibatch 10/10, validation error 48.600000 %
epoch 5, minibatch 10/10, validation error 47.800000 %
epoch 6, minibatch 10/10, validation error 48.000000 %
epoch 7, minibatch 10/10, validation error 48.000000 %
epoch 8, minibatch 10/10, validation error 48.000000 %
epoch 9, minibatch 10/10, validation error 48.200000 %
epoch 10, minibatch 10/10, validation error 48.800000 %
training @ iter =  100
epoch 11, minibatch 10/10, validation error 51.000000 %
epoch 12, minibatch 10/10, validation error 50.600000 %
epoch 13, minibatch 10/10, validation error 50.000000 %
epoch 14, minibatch 10/10, validation error 50.000000 %
epoch 15, 

The training process for function test_mlp_parity ran for 0.23m


#### 12-bit input

test_mlp_parity(learning_rate=0.65, n_bit=12,n_hiddenLayers=2,L1_reg=0.00, L2_reg=0.0001, n_epochs=5000, batch_size=100, n_hidden=100, patience=500000, verbose=False)

xuns-MacBook-Pro:src xx$ python3 hw4b.py
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")
... building the model
...training
Optimization complete.
Best validation score of 0.000000 % obtained at iteration 26290, with test performance 0.000000 %
The training process for function test_mlp_parity ran for 0.69m

#### As shown above, I can achieve 100% accuracy with both 8-bit inputs and 12-bit inputs. The configuration is documented above. The configuration is also documented above. Although I achieve 100% accuracy in both case, the 12-bit input is much harder than the 8-bit input. It is very sensitive with the learning rate and the result is slightly different between different machine. The same configuration can achieve 0% error rate on my own cpu but 0.2% on GPU, therefore I paste my result in my own computer of 12-bit input

#### Bullet 2: Deep MLP

#### 8-bit input

test_mlp_parity(learning_rate=0.02, n_bit=8,n_hiddenLayers=10,L1_reg=0.00, L2_reg=0.0001, n_epochs=5000, batch_size=100, n_hidden=5,patience=200000,verbose=False)

dyn-160-39-143-120:src xx$ python3 hw4b.py
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")
... building the model
...training
Optimization complete.
Best validation score of 0.000000 % obtained at iteration 32980, with test performance 0.000000 %
The training process for function test_mlp_parity ran for 0.38m


#### 12-bit input

The training set for 12-bit input is set to 2000

test_mlp_parity(learning_rate=0.02, n_bit=12,n_hiddenLayers=8,L1_reg=0.00, L2_reg=0.0001, n_epochs=5000, batch_size=100, n_hidden=10, patience=20000, verbose=False)

xuns-MacBook-Pro:src xx$ python3 hw4b.py
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")
... building the model
...training
Optimization complete.
Best validation score of 0.000000 % obtained at iteration 54510, with test performance 0.000000 %
The training process for function test_mlp_parity ran for 0.90m

#### As we can see from the result above, we can also achieve 0% error rate for both 8-bits input and 12-bits input.  The configuration is also shown above. For the number of hidden units, with 2-layer mlp we need 10 hidden units to achieve 0 error rate for 8-bit input and 100 hidden units 12-bit input. With the 10-layer mlp we only need 5 hidden units for 8-bit input and 10 hidden unit for 12-bit input. The number of hidden unit required for deep MLP is smaller than shallow mlp. Also, 12-bit input is still more sensitive to parameters.

#### Bullet 3: RNN

#### 8-bit input

In [1]:
from hw4b import test_rnn_parity
test_rnn_parity(**{
        'nbits':8,
        'lr': 0.2,
        'verbose': True,
        'decay': True,
        'win': 1,
        'nhidden': 5,
        'nepochs': 20
    })

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


nhidden: 5
nbits: 8
verbose: True
decay: True
win: 1
nepochs: 20
lr: 0.2
... building the model
... training
('NEW BEST: epoch', 0, 'validation error', 0.50600000000000001, 'best test test error', 0.54000000000000004)
('NEW BEST: epoch', 7, 'validation error', 0.46400000000000002, 'best test test error', 0.46000000000000002)
('NEW BEST: epoch', 11, 'validation error', 0.45800000000000002, 'best test test error', 0.48999999999999999)
('NEW BEST: epoch', 13, 'validation error', 0.434, 'best test test error', 0.44)
('NEW BEST: epoch', 14, 'validation error', 0.40999999999999998, 'best test test error', 0.42999999999999999)
('NEW BEST: epoch', 15, 'validation error', 0.40000000000000002, 'best test test error', 0.41999999999999998)
('NEW BEST: epoch', 16, 'validation error', 0.28799999999999998, 'best test test error', 0.23999999999999999)
('NEW BEST: epoch', 17, 'validation error', 0.0, 'best test test error', 0.0)
('BEST RESULT: epoch', 17, 'validation error', 0.0, 'best test error', 0.0

#### 12-bit input

In [1]:
from hw4b import test_rnn_parity
test_rnn_parity(**{
        'nbits':12,
        'lr': 1,
        'verbose': True,
        'decay': True,
        'win': 1,
        'nhidden': 10,
        'nepochs': 30
    })

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


nhidden: 10
nbits: 12
verbose: True
decay: True
win: 1
nepochs: 30
lr: 1
... building the model
... training
('NEW BEST: epoch', 0, 'validation error', 0.46800000000000003, 'best test test error', 0.51000000000000001)
('NEW BEST: epoch', 8, 'validation error', 0.0, 'best test test error', 0.0)
('BEST RESULT: epoch', 8, 'validation error', 0.0, 'best test error', 0.0)


#### As shown above I can achieve 0 error rate for 8-bit input with 5 hidden units and achieve 0 error rate for 12-bit input with 10 hidden units. The window size is set to 1 since it doesn't require this parameter for parity function learning

#### Bullet 4: LSTM

#### 8-bit input

In [1]:
from hw4b import test_lstm_parity
test_lstm_parity(**{
        'nbits':8,
        'lr': 0.1,
        'verbose': True,
        'decay': True,
        'win': 1,
        'nhidden': 1,
        'nepochs': 20,
        'savemodel': False,
    })

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


verbose: True
win: 1
savemodel: False
nepochs: 20
nhidden: 1
nbits: 8
decay: True
lr: 0.1
... building the model
... training
('NEW BEST: epoch', 0, 'validation error', 0.504, 'best test test error', 0.53000000000000003)
('NEW BEST: epoch', 3, 'validation error', 0.5, 'best test test error', 0.53000000000000003)
('NEW BEST: epoch', 4, 'validation error', 0.498, 'best test test error', 0.54000000000000004)
('NEW BEST: epoch', 5, 'validation error', 0.040000000000000001, 'best test test error', 0.029999999999999999)
('NEW BEST: epoch', 6, 'validation error', 0.0, 'best test test error', 0.0)
('BEST RESULT: epoch', 6, 'validation error', 0.0, 'best test error', 0.0)


#### 12-bit input

In [2]:
from hw4b import test_lstm_parity
test_lstm_parity(**{
        'nbits':12,
        'lr': 0.1,
        'verbose': True,
        'decay': True,
        'win': 1,
        'nhidden': 1,
        'nepochs': 20
    })

verbose: True
win: 1
savemodel: False
nepochs: 20
nhidden: 1
nbits: 12
decay: True
lr: 0.1
... building the model
... training
('NEW BEST: epoch', 0, 'validation error', 0.50600000000000001, 'best test test error', 0.51000000000000001)
('NEW BEST: epoch', 2, 'validation error', 0.48999999999999999, 'best test test error', 0.5)
('NEW BEST: epoch', 3, 'validation error', 0.47199999999999998, 'best test test error', 0.52000000000000002)
('NEW BEST: epoch', 9, 'validation error', 0.46400000000000002, 'best test test error', 0.42999999999999999)
('NEW BEST: epoch', 11, 'validation error', 0.40999999999999998, 'best test test error', 0.47999999999999998)
('NEW BEST: epoch', 12, 'validation error', 0.0, 'best test test error', 0.0)
('BEST RESULT: epoch', 12, 'validation error', 0.0, 'best test error', 0.0)


#### As shown above, I can achieve 0 error rate with lstm with only one hidden unit for both 8-bit input and 12-bit input